apps/keyboard/js/keyboard/input_method_manager.js

```diff
@@ -521,22 +474,24 @@ InputMethodManager.prototype.deactivateIMEngine = function() {
 
 InputMethodManager.prototype.handleEvent = function(evt) {
   this.app.console.info('InputMethodManager.handleEvent()', evt);
-  switch (evt.type) {
-    case 'selectionchange':
-      this.app.console.log(
-        'InputMethodManager::currentIMEngine.selectionChange()', evt.detail);
-      this.currentIMEngine.selectionChange(evt.detail);
 
-      break;
+  var inputContext = this.app.inputContext;
 
-    case 'surroundingtextchange':
-      this.app.console.log(
-        'InputMethodManager::currentIMEngine.surroundingtextChange()',
-        evt.detail);
-      this.currentIMEngine.surroundingtextChange(evt.detail);
+  var inputContextStates = {
+    // TODO: ownAction is not trustworthy; remove it one day and ask IMengines
+    // to calculate themselves.
+    ownAction: evt.detail.ownAction,
 
-      break;
-  }
+    type: inputContext.inputType,
+    inputmode: inputContext.inputMode,
+    selectionStart: inputContext.selectionStart,
+    selectionEnd: inputContext.selectionEnd,
+    text: inputContext.text
+  };
+
+  this.app.console.log(
+    'InputMethodManager::currentIMEngine.stateChange()', inputContextStates);
+  this.currentIMEngine.stateChange(inputContextStates);
```

Comment:
As I know, inputContext will be updated every time inputcontextchange eveyt triggered by mozInputMethod, so it should obtain the latest data. 

Pass the it to IME for updating state when selectionchange or surroundingtextchange triggered.